File: tests/engine/fixtures.ts
/**
 * Test fixtures: a synthetic card set and encounter factory exercising the
 * full v1.4 vocabulary without any real content (tests are content-agnostic,
 * like the engine itself).
 */
import type {
  CardDefinition,
  CombatState,
  CombatAction,
  EncounterConfig,
  InfoNugget,
} from '../../src/engine';
import { buildInitialState, reduce } from '../../src/engine';

const base = {
  keywords: [] as CardDefinition['keywords'],
  color: 'Colorless' as const,
  supertype: 'Skill' as const,
  subtype: null,
  effectText: 'test',
};

export const CARDS: Record<string, CardDefinition> = {};
function card(def: Partial<CardDefinition> & { id: string; cost: number }): void {
  CARDS[def.id] = { ...base, name: def.id, effects: [], ...def } as CardDefinition;
}

card({ id: 'ponder', cost: 1, effects: [{ type: 'DRAW_CARDS', value: 1 }] });

// Player-side skills
card({ id: 'p_noop', cost: 1 });
card({ id: 'p_free', cost: 0 });
card({ id: 'p_break', cost: 2, effects: [{ type: 'BREAK_SHIELDS', target: 'opponent', count: 1 }] });
card({ id: 'p_break3', cost: 1, effects: [{ type: 'BREAK_SHIELDS', target: 'opponent', count: 3 }] });
card({ id: 'p_gain_priority', cost: 0, effects: [{ type: 'MODIFY_PRIORITY', value: 3 }] });
card({ id: 'p_expensive', cost: 9 });
card({ id: 'p_draw', cost: 1, effects: [{ type: 'DRAW_CARDS', value: 2 }] });
card({ id: 'p_patience_up', cost: 1, effects: [{ type: 'MODIFY_PATIENCE', value: 3 }] });
card({ id: 'p_patience_down', cost: 1, effects: [{ type: 'MODIFY_PATIENCE', value: -3 }] });
card({ id: 'p_lie', cost: 0, keywords: ['Lie'] });
card({ id: 'p_safety', cost: 1, keywords: ['Safety'] });
card({
  id: 'p_shield_trigger',
  cost: 1,
  keywords: ['Shield Trigger'],
  shieldTriggerEffects: [{ type: 'MODIFY_PATIENCE', value: 2 }],
});
card({
  id: 'p_heavy',
  cost: 2,
  keywords: ['Heavy Hand'],
  effects: [{ type: 'MODIFY_PATIENCE', value: -1 }],
  heavyHandEffects: [{ type: 'MODIFY_PATIENCE', value: -1 }, { type: 'BREAK_SHIELDS', target: 'opponent', count: 1 }],
});
card({
  id: 'p_choose',
  cost: 1,
  effects: [
    { type: 'CHOOSE_NUMBER', min: 1, max: 10 },
    { type: 'MODIFY_PATIENCE', value: 1, scale: { kind: 'CHOSEN_NUMBER' } },
  ],
});
card({
  id: 'p_trap_cancel',
  cost: 2,
  keywords: ['Trap'],
  subtype: 'Trap',
  trapTrigger: { event: 'CARD_STAGED', controllerFilter: 'opponent' },
  effects: [{ type: 'CANCEL_STAGED_CARD' }],
});
card({
  id: 'p_trap_rapport',
  cost: 2,
  keywords: ['Trap', 'Rapport'],
  subtype: 'Trap',
  rapport: { min: 1, max: 10, checked: { kind: 'STAGED_CARD_COST' } },
  trapTrigger: {
    event: 'CARD_STAGED',
    controllerFilter: 'opponent',
    condition: { compare: { lhs: { kind: 'STAGED_CARD_COST' }, op: 'eq', rhs: { kind: 'CHOSEN_NUMBER' } } },
  },
  effects: [{ type: 'CANCEL_STAGED_CARD' }, { type: 'MODIFY_PATIENCE', value: 1, scale: { kind: 'CHOSEN_NUMBER' } }],
});
card({
  id: 'p_trap_on_patience',
  cost: 1,
  keywords: ['Trap'],
  subtype: 'Trap',
  trapTrigger: {
    event: 'PATIENCE_CHANGED',
    condition: { compare: { lhs: { kind: 'EVENT_NEW_VALUE' }, op: 'lt', rhs: { kind: 'CONST', value: 5 } } },
  },
  effects: [{ type: 'MODIFY_PRIORITY', value: 2 }],
});
// NPC trap that responds to PLAYER_TURN_START (a player-owned trap cannot:
// it expires at §4.1 step 4, before the step-9 dispatch).
card({
  id: 'n_trap_pts',
  cost: 1,
  keywords: ['Trap'],
  subtype: 'Trap',
  trapTrigger: { event: 'PLAYER_TURN_START' },
  effects: [
    {
      type: 'APPLY_RESTRICTION',
      restriction: { type: 'PREVENT_DRAW', target: 'opponent', expiry: { boundary: 'PLAYER_TURN_START', occurrences: 1 } },
    },
    { type: 'MODIFY_PATIENCE', value: -1 },
  ],
});
card({
  id: 'p_impression_on_break',
  cost: 2,
  subtype: 'Impression',
  triggeredAbilities: [
    {
      id: 'gain_on_break',
      trigger: { event: 'SHIELD_BROKEN', controllerFilter: 'self' },
      effects: [{ type: 'MODIFY_PRIORITY', value: 1 }],
    },
  ],
});
card({
  id: 'p_token_maker',
  cost: 1,
  effects: [{ type: 'CREATE_TOKEN', tokenDefinitionId: 'tok_chain', count: 1 }],
});
card({
  id: 'p_token_smash',
  cost: 1,
  effects: [{ type: 'DESTROY_TOKENS', count: 2 }],
});
card({
  id: 'p_replacer',
  cost: 1,
  effects: [
    {
      type: 'APPLY_REPLACEMENT',
      originalTokenId: 'tok_chain',
      replacementTokenId: 'tok_boom',
      expiry: { boundary: 'PLAYER_TURN_END', occurrences: 1 },
    },
  ],
});
card({
  id: 'p_scheduler',
  cost: 1,
  effects: [
    {
      type: 'SCHEDULE_EFFECTS',
      effects: [{ type: 'MODIFY_PRIORITY', value: 4 }],
      at: { boundary: 'PLAYER_TURN_START', occurrences: 1 },
    },
  ],
});
card({
  id: 'p_copy',
  cost: 1,
  effects: [{ type: 'COPY_FROM_NPC_DECK', count: 1, patienceCostOverride: { kind: 'CONST', value: 2 } }],
});
card({
  id: 'p_counter_feeder',
  cost: 0,
  effects: [{ type: 'INCREMENT_COUNTERS', counterName: 'devotion', targetDefinitionId: 'n_counter_impression', amount: 2 }],
});

card({
  id: 'p_trap_draw',
  cost: 1,
  keywords: ['Trap'],
  subtype: 'Trap',
  trapTrigger: { event: 'CARD_PLAYED', controllerFilter: 'opponent' },
  effects: [{ type: 'DRAW_CARDS', value: 1 }],
});
card({
  id: 'p_impression_battery',
  cost: 1,
  subtype: 'Impression',
  activatedAbilities: [
    { id: 'surge', name: 'Surge', cost: { patience: 1 }, effects: [{ type: 'MODIFY_PRIORITY', value: 6 }] },
  ],
});

// Information cards
card({ id: 'info_a', cost: 0, supertype: 'Information', nuggetId: 'nug_a' });
card({ id: 'info_b', cost: 0, supertype: 'Information', nuggetId: 'nug_b' });
card({ id: 'info_c', cost: 0, supertype: 'Information', nuggetId: 'nug_c' });

// Core-shield lore cards
card({ id: 'lore_1', cost: 0 });
card({ id: 'lore_2', cost: 0 });

// NPC cards
card({ id: 'n_noop', cost: 1 });
card({ id: 'n_free', cost: 0 });
card({ id: 'n_break', cost: 1, effects: [{ type: 'BREAK_SHIELDS', target: 'opponent', count: 1 }] });
card({ id: 'n_break2', cost: 1, effects: [{ type: 'BREAK_SHIELDS', target: 'opponent', count: 2 }] });
card({ id: 'n_break5', cost: 1, effects: [{ type: 'BREAK_SHIELDS', target: 'opponent', count: 5 }] });
card({ id: 'n_patience_drain', cost: 1, effects: [{ type: 'MODIFY_PATIENCE', value: -2 }] });
card({ id: 'n_guards_up', cost: 1, effects: [{ type: 'PLACE_SHIELDS', target: 'self', count: 2 }] });
card({ id: 'n_self_break', cost: 1, effects: [{ type: 'BREAK_SHIELDS', target: 'self', count: 1 }] });
card({ id: 'n_draw_blocker', cost: 1, effects: [
  {
    type: 'APPLY_RESTRICTION',
    restriction: { type: 'PREVENT_DRAW', target: 'opponent', expiry: { boundary: 'PLAYER_TURN_END', occurrences: 1 } },
  },
] });
card({
  id: 'n_impression_ts',
  cost: 1,
  subtype: 'Impression',
  turnStartEffects: [{ type: 'MODIFY_PATIENCE', value: -1 }],
});
card({
  id: 'n_counter_impression',
  cost: 0,
  subtype: 'Impression',
  thresholds: [
    { counterName: 'devotion', value: 4, consume: true, effects: [{ type: 'BREAK_SHIELDS', target: 'opponent', count: 1 }], checkPoint: 'AFTER_ANY_PLAY' },
  ],
});
card({
  id: 'n_trap_on_play',
  cost: 1,
  keywords: ['Trap'],
  subtype: 'Trap',
  trapTrigger: { event: 'CARD_PLAYED', controllerFilter: 'opponent' },
  effects: [{ type: 'MODIFY_PATIENCE', value: -1 }],
});

export const TOKENS: Record<string, CardDefinition> = {
  tok_chain: {
    ...base,
    id: 'tok_chain',
    name: 'tok_chain',
    cost: 0,
    subtype: 'Token',
    effects: [],
    leaveTriggerEffects: [{ type: 'DRAW_CARDS', value: 1 }],
  },
  tok_boom: {
    ...base,
    id: 'tok_boom',
    name: 'tok_boom',
    cost: 0,
    subtype: 'Token',
    effects: [],
    leaveTriggerEffects: [{ type: 'BREAK_SHIELDS', target: 'opponent', count: 1 }],
  },
};
// Token defs must also be resolvable through the card registry used by getDef.
for (const t of Object.values(TOKENS)) CARDS[t.id] = t;

export const NUGGETS: Record<string, InfoNugget> = {
  nug_a: { id: 'nug_a', name: 'Nugget A', description: 'a' },
  nug_b: { id: 'nug_b', name: 'Nugget B', description: 'b' },
  nug_c: { id: 'nug_c', name: 'Nugget C', description: 'c' },
};

export function makeEncounter(overrides: Partial<EncounterConfig> = {}): EncounterConfig {
  return {
    id: 'test',
    displayName: 'Test Encounter',
    minTurnStartPriority: 3,
    firstTurnBonusPriority: 2,
    maxPriority: 10,
    startingSide: 'player',
    opponentPatience: 10,
    npcGuardShieldCount: 2,
    opponentShields: [
      { cardId: 'lore_1', isHint: false, loreDescription: 'lore one', keyNuggetIds: ['nug_a'] },
      { cardId: 'lore_2', isHint: true, hintText: 'hint', loreDescription: 'lore two', keyNuggetIds: ['nug_b'] },
    ],
    npcHandLimit: 5,
    playerDummyShieldSlots: 3,
    allowedCoreShields: [],
    nuggetOverrides: [
      { nuggetId: 'nug_a', cost: 1, effects: [{ type: 'DRAW_CARDS', value: 1 }], effectText: 'draw 1' },
      { nuggetId: 'nug_b', cost: 0, effects: [], effectText: 'nothing' },
    ],
    traits: [],
    enemyDeckCardIds: ['n_noop', 'n_noop', 'n_noop', 'n_noop', 'n_noop', 'n_noop'],
    lieThreshold: 2,
    ...overrides,
  };
}

export interface StartOptions {
  config?: Partial<EncounterConfig>;
  deck?: string[];
  collection?: string[];
  seed?: number;
  persistent?: Parameters<typeof buildInitialState>[0]['persistent'];
}

export function start(opts: StartOptions = {}): CombatState {
  return buildInitialState({
    config: makeEncounter(opts.config),
    cards: CARDS,
    tokens: TOKENS,
    nuggets: NUGGETS,
    playerDeckCardIds: opts.deck ?? Array(12).fill('p_noop'),
    collectionCardIds: opts.collection ?? [],
    seed: opts.seed ?? 42,
    persistent: opts.persistent,
  });
}

/** Play the hand card with the given definition id (must be in hand). */
export function playCard(state: CombatState, defId: string, heavyHand = false): CombatState {
  const idx = state.player.hand.findIndex((c) => c.definitionId === defId);
  if (idx === -1) throw new Error(`${defId} not in hand: ${state.player.hand.map((c) => c.definitionId).join(',')}`);
  return reduce(state, { type: 'PLAY_CARD', handIndex: idx, heavyHand });
}

export function act(state: CombatState, action: CombatAction): CombatState {
  return reduce(state, action);
}

/** End the player turn, auto-completing BotM with an empty keep. */
export function endTurn(state: CombatState): CombatState {
  let s = reduce(state, { type: 'END_TURN' });
  if (s.phase === 'BotMSelect') s = reduce(s, { type: 'BOTM_SELECT', keepHandIndices: [] });
  return s;
}

/** Run the whole NPC turn via ADVANCE until control returns to the player or the game ends. */
export function runNpcTurn(state: CombatState): CombatState {
  let s = state;
  let guard = 0;
  while (s.phase === 'EnemyPending') {
    if (++guard > 50) throw new Error('NPC turn did not terminate');
    s = reduce(s, { type: 'ADVANCE' });
    while (s.pendingBlock?.type === 'reveal' || s.pendingBlock?.type === 'deckReveal') {
      s = reduce(s, { type: 'ACKNOWLEDGE' });
    }
  }
  return s;
}

File: csharp-engine/Breakthrough.Engine.Tests/Fixtures.cs
// Test fixtures: a synthetic card set and encounter factory exercising the
// full v1.4 vocabulary without any real content (tests are content-agnostic,
// like the engine itself).
//
// 1:1 C# port of tests/engine/fixtures.ts.

using Breakthrough.Engine;
using Breakthrough.Engine.Json;

namespace Breakthrough.Engine.Tests;

public static class Fixtures
{
    public static readonly Dictionary<string, CardDefinition> CARDS = BuildCards();
    public static readonly Dictionary<string, CardDefinition> TOKENS = BuildTokens();
    public static readonly Dictionary<string, InfoNugget> NUGGETS = new()
    {
        ["nug_a"] = new InfoNugget("nug_a", "Nugget A", "a"),
        ["nug_b"] = new InfoNugget("nug_b", "Nugget B", "b"),
        ["nug_c"] = new InfoNugget("nug_c", "Nugget C", "c"),
    };

    private static CardDefinition Card(
        string id,
        int cost,
        IReadOnlyList<Effect>? effects = null,
        IReadOnlyList<string>? keywords = null,
        string supertype = Supertypes.Skill,
        string? subtype = null,
        string? nuggetId = null,
        IReadOnlyList<Effect>? shieldTriggerEffects = null,
        IReadOnlyList<Effect>? heavyHandEffects = null,
        TriggerCondition? trapTrigger = null,
        RapportConfig? rapport = null,
        IReadOnlyList<TriggeredAbility>? triggeredAbilities = null,
        IReadOnlyList<ActivatedAbility>? activatedAbilities = null,
        IReadOnlyList<Effect>? turnStartEffects = null,
        IReadOnlyList<ThresholdDef>? thresholds = null,
        IReadOnlyList<Effect>? leaveTriggerEffects = null) => new()
    {
        Id = id,
        Name = id,
        Cost = cost,
        Color = "Colorless",
        Supertype = supertype,
        Subtype = subtype,
        Keywords = keywords ?? [],
        Effects = effects ?? [],
        EffectText = "test",
        NuggetId = nuggetId,
        ShieldTriggerEffects = shieldTriggerEffects,
        HeavyHandEffects = heavyHandEffects,
        TrapTrigger = trapTrigger,
        Rapport = rapport,
        TriggeredAbilities = triggeredAbilities,
        ActivatedAbilities = activatedAbilities,
        TurnStartEffects = turnStartEffects,
        Thresholds = thresholds,
        LeaveTriggerEffects = leaveTriggerEffects,
    };

    private static Dictionary<string, CardDefinition> BuildCards()
    {
        var cards = new List<CardDefinition>
        {
            Card("ponder", 1, effects: [new DrawCardsEffect(1)]),

            // Player-side skills
            Card("p_noop", 1),
            Card("p_free", 0),
            Card("p_break", 2, effects: [new BreakShieldsEffect(RelSide.Opponent, 1)]),
            Card("p_break3", 1, effects: [new BreakShieldsEffect(RelSide.Opponent, 3)]),
            Card("p_gain_priority", 0, effects: [new ModifyPriorityEffect(3)]),
            Card("p_expensive", 9),
            Card("p_draw", 1, effects: [new DrawCardsEffect(2)]),
            Card("p_patience_up", 1, effects: [new ModifyPatienceEffect(3)]),
            Card("p_patience_down", 1, effects: [new ModifyPatienceEffect(-3)]),
            Card("p_lie", 0, keywords: [Keywords.Lie]),
            Card("p_safety", 1, keywords: [Keywords.Safety]),
            Card("p_shield_trigger", 1,
                keywords: [Keywords.ShieldTrigger],
                shieldTriggerEffects: [new ModifyPatienceEffect(2)]),
            Card("p_heavy", 2,
                keywords: [Keywords.HeavyHand],
                effects: [new ModifyPatienceEffect(-1)],
                heavyHandEffects: [new ModifyPatienceEffect(-1), new BreakShieldsEffect(RelSide.Opponent, 1)]),
            Card("p_choose", 1, effects:
            [
                new ChooseNumberEffect(1, 10),
                new ModifyPatienceEffect(1) { Scale = new ChosenNumberQ() },
            ]),
            Card("p_trap_cancel", 2,
                keywords: [Keywords.Trap],
                subtype: Subtypes.Trap,
                trapTrigger: new TriggerCondition(EventTypes.CardStaged) { ControllerFilter = "opponent" },
                effects: [new CancelStagedCardEffect()]),
            Card("p_trap_rapport", 2,
                keywords: [Keywords.Trap, Keywords.Rapport],
                subtype: Subtypes.Trap,
                rapport: new RapportConfig(1, 10, new StagedCardCostQ()),
                trapTrigger: new TriggerCondition(EventTypes.CardStaged)
                {
                    ControllerFilter = "opponent",
                    Condition = new CompareCondition(new StagedCardCostQ(), Comparators.Eq, new ChosenNumberQ()),
                },
                effects: [new CancelStagedCardEffect(), new ModifyPatienceEffect(1) { Scale = new ChosenNumberQ() }]),
            Card("p_trap_on_patience", 1,
                keywords: [Keywords.Trap],
                subtype: Subtypes.Trap,
                trapTrigger: new TriggerCondition(EventTypes.PatienceChanged)
                {
                    Condition = new CompareCondition(new EventNewValueQ(), Comparators.Lt, new ConstQ(5)),
                },
                effects: [new ModifyPriorityEffect(2)]),
            // NPC trap that responds to PLAYER_TURN_START (a player-owned trap
            // cannot: it expires at §4.1 step 4, before the step-9 dispatch).
            Card("n_trap_pts", 1,
                keywords: [Keywords.Trap],
                subtype: Subtypes.Trap,
                trapTrigger: new TriggerCondition(EventTypes.PlayerTurnStart),
                effects:
                [
                    new ApplyRestrictionEffect(new RestrictionDef(RestrictionTypes.PreventDraw, RelTargets.Opponent)
                    {
                        Expiry = new BoundaryRef { Boundary = BoundaryNames.PlayerTurnStart, Occurrences = 1 },
                    }),
                    new ModifyPatienceEffect(-1),
                ]),
            Card("p_impression_on_break", 2,
                subtype: Subtypes.Impression,
                triggeredAbilities:
                [
                    new TriggeredAbility("gain_on_break",
                        new TriggerCondition(EventTypes.ShieldBroken) { ControllerFilter = "self" },
                        [new ModifyPriorityEffect(1)]),
                ]),
            Card("p_token_maker", 1, effects: [new CreateTokenEffect("tok_chain", 1)]),
            Card("p_token_smash", 1, effects: [new DestroyTokensEffect(2)]),
            Card("p_replacer", 1, effects:
            [
                new ApplyReplacementEffect("tok_chain", "tok_boom")
                {
                    Expiry = new BoundaryRef { Boundary = BoundaryNames.PlayerTurnEnd, Occurrences = 1 },
                },
            ]),
            Card("p_scheduler", 1, effects:
            [
                new ScheduleEffectsEffect(
                    [new ModifyPriorityEffect(4)],
                    new BoundaryRef { Boundary = BoundaryNames.PlayerTurnStart, Occurrences = 1 }),
            ]),
            Card("p_copy", 1, effects:
            [
                new CopyFromNpcDeckEffect(1) { PatienceCostOverride = new ConstQ(2) },
            ]),
            Card("p_counter_feeder", 0, effects:
            [
                new IncrementCountersEffect("devotion", "n_counter_impression", 2),
            ]),

            Card("p_trap_draw", 1,
                keywords: [Keywords.Trap],
                subtype: Subtypes.Trap,
                trapTrigger: new TriggerCondition(EventTypes.CardPlayed) { ControllerFilter = "opponent" },
                effects: [new DrawCardsEffect(1)]),
            Card("p_impression_battery", 1,
                subtype: Subtypes.Impression,
                activatedAbilities:
                [
                    new ActivatedAbility("surge", "Surge",
                        new ActivatedAbilityCost { Patience = 1 },
                        [new ModifyPriorityEffect(6)]),
                ]),

            // Information cards
            Card("info_a", 0, supertype: Supertypes.Information, nuggetId: "nug_a"),
            Card("info_b", 0, supertype: Supertypes.Information, nuggetId: "nug_b"),
            Card("info_c", 0, supertype: Supertypes.Information, nuggetId: "nug_c"),

            // Core-shield lore cards
            Card("lore_1", 0),
            Card("lore_2", 0),

            // NPC cards
            Card("n_noop", 1),
            Card("n_free", 0),
            Card("n_break", 1, effects: [new BreakShieldsEffect(RelSide.Opponent, 1)]),
            Card("n_break2", 1, effects: [new BreakShieldsEffect(RelSide.Opponent, 2)]),
            Card("n_break5", 1, effects: [new BreakShieldsEffect(RelSide.Opponent, 5)]),
            Card("n_patience_drain", 1, effects: [new ModifyPatienceEffect(-2)]),
            Card("n_guards_up", 1, effects: [new PlaceShieldsEffect(2)]),
            Card("n_self_break", 1, effects: [new BreakShieldsEffect(RelSide.Self, 1)]),
            Card("n_draw_blocker", 1, effects:
            [
                new ApplyRestrictionEffect(new RestrictionDef(RestrictionTypes.PreventDraw, RelTargets.Opponent)
                {
                    Expiry = new BoundaryRef { Boundary = BoundaryNames.PlayerTurnEnd, Occurrences = 1 },
                }),
            ]),
            Card("n_impression_ts", 1,
                subtype: Subtypes.Impression,
                turnStartEffects: [new ModifyPatienceEffect(-1)]),
            Card("n_counter_impression", 0,
                subtype: Subtypes.Impression,
                thresholds:
                [
                    new ThresholdDef("devotion", 4, true, [new BreakShieldsEffect(RelSide.Opponent, 1)])
                    {
                        CheckPoint = CheckPoints.AfterAnyPlay,
                    },
                ]),
            Card("n_trap_on_play", 1,
                keywords: [Keywords.Trap],
                subtype: Subtypes.Trap,
                trapTrigger: new TriggerCondition(EventTypes.CardPlayed) { ControllerFilter = "opponent" },
                effects: [new ModifyPatienceEffect(-1)]),
        };
        var dict = cards.ToDictionary(c => c.Id);
        // Token defs must also be resolvable through the card registry used by GetDef.
        foreach (var t in BuildTokens().Values) dict[t.Id] = t;
        return dict;
    }

    private static Dictionary<string, CardDefinition> BuildTokens() => new()
    {
        ["tok_chain"] = Card("tok_chain", 0,
            subtype: Subtypes.Token,
            leaveTriggerEffects: [new DrawCardsEffect(1)]),
        ["tok_boom"] = Card("tok_boom", 0,
            subtype: Subtypes.Token,
            leaveTriggerEffects: [new BreakShieldsEffect(RelSide.Opponent, 1)]),
    };

    public static EncounterConfig MakeEncounter(Action<EncounterConfig>? overrides = null)
    {
        var config = new EncounterConfig
        {
            Id = "test",
            DisplayName = "Test Encounter",
            MinTurnStartPriority = 3,
            FirstTurnBonusPriority = 2,
            MaxPriority = 10,
            StartingSide = Side.Player,
            OpponentPatience = 10,
            NpcGuardShieldCount = 2,
            OpponentShields =
            [
                new NpcCoreShieldDef { CardId = "lore_1", IsHint = false, LoreDescription = "lore one", KeyNuggetIds = ["nug_a"] },
                new NpcCoreShieldDef { CardId = "lore_2", IsHint = true, HintText = "hint", LoreDescription = "lore two", KeyNuggetIds = ["nug_b"] },
            ],
            NpcHandLimit = 5,
            PlayerDummyShieldSlots = 3,
            AllowedCoreShields = [],
            NuggetOverrides =
            [
                new NuggetOverride("nug_a", 1, [new DrawCardsEffect(1)], "draw 1"),
                new NuggetOverride("nug_b", 0, [], "nothing"),
            ],
            Traits = [],
            EnemyDeckCardIds = ["n_noop", "n_noop", "n_noop", "n_noop", "n_noop", "n_noop"],
            LieThreshold = 2,
        };
        overrides?.Invoke(config);
        return config;
    }

    public sealed class StartOptions
    {
        public Action<EncounterConfig>? Config { get; init; }
        public IReadOnlyList<string>? Deck { get; init; }
        public IReadOnlyList<string>? Collection { get; init; }
        public int Seed { get; init; } = 42;
        public PersistentInput? Persistent { get; init; }
    }

    public static CombatState Start(StartOptions? opts = null)
    {
        opts ??= new StartOptions();
        return Setup.BuildInitialState(new SetupInput
        {
            Config = MakeEncounter(opts.Config),
            Cards = CARDS,
            Tokens = TOKENS,
            Nuggets = NUGGETS,
            PlayerDeckCardIds = opts.Deck ?? Enumerable.Repeat("p_noop", 12).ToList(),
            CollectionCardIds = opts.Collection ?? [],
            Seed = opts.Seed,
            Persistent = opts.Persistent,
        });
    }

    /// <summary>Play the hand card with the given definition id (must be in hand).</summary>
    public static CombatState PlayCardByDef(CombatState state, string defId, bool heavyHand = false)
    {
        int idx = state.Player.Hand.FindIndex(c => c.DefinitionId == defId);
        if (idx == -1)
            throw new InvalidOperationException(
                $"{defId} not in hand: {string.Join(",", state.Player.Hand.Select(c => c.DefinitionId))}");
        return Reducer.Reduce(state, new PlayCard(idx, heavyHand));
    }

    public static CombatState Act(CombatState state, CombatAction action) => Reducer.Reduce(state, action);

    /// <summary>End the player turn, auto-completing BotM with an empty keep.</summary>
    public static CombatState EndPlayerTurn(CombatState state)
    {
        var s = Reducer.Reduce(state, new EndTurn());
        if (s.Phase == Phases.BotMSelect) s = Reducer.Reduce(s, new BotmSelect([]));
        return s;
    }

    /// <summary>Run the whole NPC turn via ADVANCE until control returns to the player or the game ends.</summary>
    public static CombatState RunNpcTurn(CombatState state)
    {
        var s = state;
        int guard = 0;
        while (s.Phase == Phases.EnemyPending)
        {
            if (++guard > 50) throw new InvalidOperationException("NPC turn did not terminate");
            s = Reducer.Reduce(s, new Advance());
            while (s.PendingBlock is RevealBlock or DeckRevealBlock)
            {
                s = Reducer.Reduce(s, new Acknowledge());
            }
        }
        return s;
    }

    // ── Assertion helpers (JSON.stringify / log-query equivalents) ───────────

    /// <summary>Canonical JSON of the full state (the tests' JSON.stringify(s) equivalent).</summary>
    public static string StateJson(CombatState s) => EngineJson.Serialize(s);

    /// <summary>Canonical JSON of any engine value (the tests' JSON.stringify(x) equivalent).</summary>
    public static string ToJson<T>(T value) => EngineJson.Serialize(value);

    /// <summary>JSON of the state with the log ignored ({ ...s, log: null, logSeq: 0 }).</summary>
    public static string StateJsonIgnoringLog(CombatState s)
    {
        var clone = s.Clone();
        clone.Log = [];
        clone.LogSeq = 0;
        return EngineJson.Serialize(clone);
    }

    public static LogEntry? LastLog(CombatState s) => s.Log.Count > 0 ? s.Log[^1] : null;

    public static bool HasLog(CombatState s, string type) => s.Log.Any(l => l.Type == type);

    public static object? Data(LogEntry? entry, string key) =>
        entry?.Data != null && entry.Data.TryGetValue(key, out var v) ? v : null;

    public static bool DataEquals(LogEntry? entry, string key, object? expected) => Equals(Data(entry, key), expected);
}
